js/plugins/vertexai/src/vector-search/retrievers.ts

```diff
@@ -0,0 +1,162 @@
+  const vectorSearchOptions = params.pluginOptions.vectorSearchIndexOptions;
+  const defaultEmbedder = params.defaultEmbedder;
+
+  const retrievers: RetrieverAction<z.ZodTypeAny>[] = [];
+
+  if (!vectorSearchOptions || vectorSearchOptions.length === 0) {
+    return retrievers;
+  }
+
+  for (const vectorSearchOption of vectorSearchOptions) {
+    const { documentRetriever, indexId, publicDomainName } = vectorSearchOption;
+    const embedder = vectorSearchOption.embedder ?? defaultEmbedder;
+    const embedderOptions = vectorSearchOption.embedderOptions;
+
+    const retriever = defineRetriever(
+      {
+        name: `vertexai/${indexId}`,
+        configSchema: VertexAIVectorRetrieverOptionsSchema,
+      },
+      async (content, options) => {
+        const queryEmbeddings = await embed({
+          embedder,
+          options: embedderOptions,
+          content,
+        });
+
+        const accessToken = await params.authClient.getAccessToken();
+
+        if (!accessToken) {
+          throw new Error(
+            'Access token is required to define Vertex AI retriever'
+          );
+        }
+
+        const projectId = params.pluginOptions.projectId;
+        if (!projectId) {
+          throw new Error(
+            'Project ID is required to define Vertex AI retriever'
+          );
+        }
+        const projectNumber = await getProjectNumber(projectId);
+        const location = params.pluginOptions.location;
+        if (!location) {
+          throw new Error('Location is required to define Vertex AI retriever');
+        }
+
+        try {
+          let res = await queryPublicEndpoint({
+            featureVector: queryEmbeddings,
+            neighborCount: options?.k || DEFAULT_K,
+            accessToken,
+            projectId,
+            location,
+            publicDomainName,
+            projectNumber,
+            indexEndpointId: vectorSearchOption.indexEndpointId,
+            deployedIndexId: vectorSearchOption.deployedIndexId,
+          });
+          const nearestNeighbors = res.nearestNeighbors;
+
+          const queryRes = nearestNeighbors ? nearestNeighbors[0] : null;
+          const neighbors = queryRes ? queryRes.neighbors : null;
+          if (!nearestNeighbors || !queryRes || !neighbors) {
+            return { documents: [] };
+          }
+
+          const documents = await documentRetriever(neighbors, options);
+
+          return { documents };
+        } catch (error) {
+          handleRetrieverError(error, indexId);
+        }
+      }
+    );
+
+    retrievers.push(retriever);
+  }
+
+  return retrievers;
+}
+
+/**
+ * Handles errors that occur during the retriever process.
+ *
+ * @param {unknown} error - The error that occurred.
+ * @param {string} indexId - The ID of the index being queried.
+ * @throws {Error} - Throws the error after logging it.
+ */
+function handleRetrieverError(error: unknown, indexId: string): never {
+  if (error instanceof Error) {
```

Comment:
just curious -- when would this be false?

Reply:
yeah I can remove this tbh, It always bugs me about ts error handling that anything can be thrown